extras/ai-battle/HeadlessGame.cpp

```diff
@@ -0,0 +1,264 @@
+using bfs::canonical;
+
+#ifdef WIN32
+HANDLE setupStdOut();
+void printConsole(const char* fmt, ...);
+#elif defined(__GNUC__)
+void printConsole(const char* fmt, ...) __attribute__((format(printf, 1, 2)));
+#else
+void printConsole(const char* fmt, ...);
+#endif
+
+HeadlessGame::HeadlessGame(const GlobalGameSettings& ggs, const bfs::path& map, const std::vector<AI::Info>& ais)
+    : map_(map), game_(ggs, std::make_unique<EventManager>(0), GeneratePlayerInfo(ais)), world_(game_.world_),
+      em_(*static_cast<EventManager*>(game_.em_.get()))
+{
+    MapLoader loader(world_);
+    if(!loader.Load(map))
+        throw std::runtime_error("Could not load " + map.string());
+
+    players_.clear();
+    for(unsigned playerId = 0; playerId < world_.GetNumPlayers(); ++playerId)
+        players_.push_back(AIFactory::Create(world_.GetPlayer(playerId).aiInfo, playerId, world_));
+
+    world_.InitAfterLoad();
+}
+
+HeadlessGame::~HeadlessGame()
+{
+    Close();
+}
+
+void HeadlessGame::Run(unsigned maxGF)
+{
+    AsyncChecksum checksum;
+    gameStartTime_ = std::chrono::steady_clock::now();
+    auto nextReport = gameStartTime_ + std::chrono::seconds(1);
+
+    game_.Start(false);
+
+    while(em_.GetCurrentGF() < maxGF && !game_.IsGameFinished())
+    {
+        // In the actual game, the network frame intervall is based on ping (highest_ping < NFW-length < 20*gf_length).
+        bool isnfw = em_.GetCurrentGF() % 20 == 0;
+
+        if(isnfw)
+        {
+            if(replay_.IsRecording())
+                checksum = AsyncChecksum::create(game_);
+            for(unsigned playerId = 0; playerId < world_.GetNumPlayers(); ++playerId)
+            {
+                world_.GetPlayer(playerId);
+                AIPlayer* player = players_[playerId].get();
+                PlayerGameCommands cmds;
+                cmds.gcs = player->FetchGameCommands();
+
+                if(replay_.IsRecording() && !cmds.gcs.empty())
+                {
+                    cmds.checksum = checksum;
+                    replay_.AddGameCommand(em_.GetCurrentGF(), playerId, cmds);
+                }
+
+                for(const gc::GameCommandPtr& gc : cmds.gcs)
+                    gc->Execute(world_, player->GetPlayerId());
+            }
+        }
+
+        for(auto& player : players_)
+            player->RunGF(em_.GetCurrentGF(), isnfw);
+
+        game_.RunGF();
+
+        if(replay_.IsRecording())
+            replay_.UpdateLastGF(em_.GetCurrentGF());
+
+        if(std::chrono::steady_clock::now() > nextReport)
+        {
+            nextReport += std::chrono::seconds(1);
+            PrintState();
+        }
+    }
+    PrintState();
+}
+
+void HeadlessGame::Close()
+{
+    bnw::cout << '\n';
+
+    if(replay_.IsRecording())
+    {
+        replay_.StopRecording();
+        bnw::cout << "Replay written to " << canonical(replayPath_) << '\n';
+    }
+
+    replay_.Close();
+}
+
+void HeadlessGame::StartReplay(const bfs::path& path, unsigned random_init)
```

Comment:
Better name as I thought this was for starting *from* a replay

```suggestion
void HeadlessGame::RecordReplay(const bfs::path& path, unsigned random_init)
```